database/scrapers/event_calendar.ts

```diff
@@ -0,0 +1,230 @@
+    private task: TaskHandle,
+  ) {
+    if (
+      !this.calendarString.startsWith("BEGIN:VCALENDAR") ||
+      !this.calendarString.trimEnd().endsWith("END:VCALENDAR")
+    ) {
+      throw new Error("Calendar invalid");
+    }
+    const calendarRoot = parseICS(calendarString, (message) =>
+      task.update(message),
+    ).VCALENDAR;
+    if (calendarRoot === undefined) {
+      throw new Error("Calendar invalid - missing calendar root");
+    }
+    const eventArray = (calendarRoot as ICSRoot).VEVENT;
+    if (eventArray === undefined || !Array.isArray(eventArray)) {
+      throw new Error("Calendar invalid - missing event array");
+    }
+    this.events = eventArray;
+  }
+
+  public getLastEvent(): GoogleCalendarEventDto | null {
+    let event = this.extractLastEvent();
+    while (event === undefined) {
+      event = this.extractLastEvent();
+    }
+    return event;
+  }
+
+  //null - none left, undefined - malformed event
+  private extractLastEvent(): GoogleCalendarEventDto | null | undefined {
+    if (this.events.length === 0) {
+      return null;
+    }
+    const eventBlock = this.events.pop();
+    if (eventBlock === undefined) {
+      return undefined;
+    }
+    if (typeof eventBlock === "string" || Array.isArray(eventBlock)) {
+      this.task.update(
+        "Invalid calendar event - block is not an ICSObject. Skipping...",
+      );
+      return undefined;
+    }
+    return this.parseEvent(eventBlock);
+  }
+
+  private static assertValidDateTime(
+    dateValue: ICSValue | undefined,
+  ): DateTime<true> | null {
+    if (dateValue === undefined || Array.isArray(dateValue)) {
+      return null;
+    }
+    if (typeof dateValue === "string") {
+      //timestamp: dateValue = 20180424T00:13:21
+      const dateTime = DateTime.fromISO(dateValue);
+      if (dateTime.isValid) {
+        return dateTime;
+      }
+      return null;
+    }
+    const nestedDate = dateValue.VALUE; //full day date:  dateValue = { "VALUE": "DATE:20180421" }
+    if (
+      nestedDate === undefined ||
+      typeof nestedDate !== "string" ||
+      nestedDate.length !== 13
+    ) {
+      return null;
+    }
+    const dateTime = DateTime.fromISO(nestedDate.substring(5));
+    if (dateTime.isValid) {
+      return dateTime;
+    }
+    return null;
+  }
+
+  private static assertString(value: ICSValue | undefined): string | null {
+    if (value === undefined || typeof value !== "string") {
+      return null;
+    }
+    return value;
+  }
+
+  private parseEvent(eventBlock: ICSRoot): GoogleCalendarEventDto | undefined {
+    const startTime = CalendarParser.assertValidDateTime(
+      eventBlock[FIELD_LABELS.startDate],
+    );
+    const endTime = CalendarParser.assertValidDateTime(
+      eventBlock[FIELD_LABELS.endDate],
+    );
+    if (startTime === null || endTime === null) {
+      this.task.update(
+        `Invalid calendar event ${this.events.length + 1}, missing date labels. Skipping...`,
+      );
```

Comment:
the event index will definitely get out of sync here once you skip one event.
(also I accidentally made the indexes in errors/warnings from `parseICS` start at 1 :skull:)

Reply:
oh right, i forgot that's the event queue, not the list of interpreted events.